src/main/kotlin/org/rust/ide/surroundWith/expression/RustExpressionSurrounderBase.kt

```diff
@@ -20,7 +24,15 @@ abstract class RustExpressionSurrounderBase : Surrounder {
     }
 
     final override fun surroundElements(project: Project, editor: Editor, elements: Array<out PsiElement>): TextRange? {
-        return surroundExpression(project, editor, requireNotNull(targetExpr(elements)))
+        val expression = requireNotNull(targetExpr(elements))
+        val templateExpr = RustElementFactory.createExpression(project, getTemplate()) as RustExprElement
+
+        getLeafExpression(templateExpr).replace(expression)
+        val newExpression = expression.replace(templateExpr)
+
+        CodeInsightUtilCore.forcePsiPostprocessAndRestoreElement(newExpression)
+
+        return getSelectionRange(newExpression)
```

Comment:
In this case, I think this refactoring is good. When (and if :) ) we need a more low level API, we can introduce another base class, or just use a plain old `Surrounder`. 

But I think that we can do better, if we make surrounder generic over the the resulting exression

```
class RustExpressionSurrounder<E: RustExprElement> {
      abstract fun createTemplate(): R 
      abstract fun getLeafExpression(expr: R)
}
```